src/logic/map_objects/tribes/ship.cc

```diff
@@ -747,6 +776,61 @@ bool Ship::ship_update_expedition(Game& game, Bob::State& /* state */) {
 	return false;  // Continue with the regular expedition updates
 }
 
+void Ship::remember_detected_portspace(const Coords& coords) {
+	const EditorGameBase& egbase = owner().egbase();
+	const Map& map = egbase.map();
+	PlayerNumber space_owner = map[coords].get_owned_by();
+
+	if (DetectedPortSpace* dps = get_owner()->has_detected_port_space(coords); dps != nullptr) {
+		dps->owner = space_owner;
+		return;
+	}
+
+	std::unique_ptr<DetectedPortSpace> dps(new DetectedPortSpace());
+	dps->coords = coords;
+	dps->owner = space_owner;
+	dps->time_discovered = egbase.get_gametime();
+	dps->discovering_ship = get_shipname();
+	dps->dockpoints = map.find_portdock(coords, true);
+
+	Coords nearest_dock = Coords::null();
+	uint32_t nearest_distance = std::numeric_limits<uint32_t>::max();
+
+	for (const auto& warehouse : owner().get_building_statistics(owner().tribe().port())) {
+		if (warehouse.is_constructionsite) {
+			continue;
+		}
+		uint32_t d = map.calc_distance(coords, warehouse.pos);
+		if (d < nearest_distance) {
+			nearest_distance = d;
+			nearest_dock = warehouse.pos;
+		}
+	}
+
+	dps->direction_from_portdock = WalkingDir::IDLE;
+	if (static_cast<bool>(nearest_dock)) {
+		dps->nearest_portdock =
+		   dynamic_cast<const Warehouse&>(*map[nearest_dock].get_immovable()).get_warehouse_name();
+		// Find the main direction from the nearest own port to the portspace.
+		Path path;
+		map.findpath(nearest_dock, coords, 0, path, CheckStepAlwaysTrue());
+		std::array<unsigned, LAST_DIRECTION> counter{};
+		unsigned highest = 0;
+		for (size_t i = 0; i < path.get_nsteps(); ++i) {
+			counter[path[i] - 1]++;
+			highest = std::max(highest, counter[path[i] - 1]);
+		}
+		for (uint8_t i = 0; i < LAST_DIRECTION; ++i) {
+			if (counter[i] == highest) {
+				dps->direction_from_portdock = static_cast<WalkingDir>(i + 1);
+				break;
+			}
+		}
+	}
```

Comment:
Shouldn't we just calculate the direction from the difference of the coordinates? (I know odd row difference needs an extra step)

Also at least North and South should also be possible, and I don't know whether we should report actual (minimap) direction, or keep approximating the diagonal directions to the walking dirs (60° on the minimap).

I can do the necessary math if you like.

Reply:
That would be great, though I'm not sure how this would work regarding map wraparound

Reply:
PR is up: https://github.com/Noordfrees/widelands/pull/17

I ended up using mapview/walkingdir angles, because actual 3D directions didn't look right in the editor.